### PR TITLE
[MPAS-Framework] New fill value mask feature

### PR DESCRIPTION
**Pull request by cbegeman**
This PR is in support of CF-compliant output from MPAS components. It adds a new optional feature to fill variables in output streams with the `missing_value` defined in the Registry everywhere a masking array, designated with its registry name by a new registry attribute `missing_value_mask`, is equal to zero. These attributes can be added either to `var`s or to `vararray`s; when added to a `vararray`, masking is applied to each `var` in the `vararray`.

The masking is only applied to variables who (a) have a `missing_value` and `missing_value_mask` attribute and (b) are members of a stream that has the attribute `useMissingValMask='true'`. `useMissingValMask='false'` by default. Thus, changes can be made to Registry attributes without affecting the default behavior of streams.

Masking is supported for Real fields with dimensions 1-5 when the masking array is an integer array with dimensions 1-3 _and_ the shape of the masking array matches up to 3 contiguous dimensions of the Real field. 

The structure of code changes is as follows:

1. `gen_inc.c` adds `missing_value_mask` as specified in the Registry as an attribute of `var`s and `vararray`s and stores the value of `missing_value_mask` in the real field type structure as `maskName`. 
2. `mpas_bootstrap_framework_phase2` calls `mpas_mask_assign_pointer`, which assigns a pointer from the real field structure to the masking array. The name of the pointer is `missingValueMaskXD` where X is the dimension of the masking array. 
3. `xml_stream_parser.c` ensures that `useMissingValMask` is added as an attribute of the stream. `stream_mgr_create_stream_c` converts the string value of `useMissingValMask` given in the streams xml file to the logical value of `useMissingValMask` in the stream type structure.
4. `MPAS_writeStream` applies the masking if the command line argument `useMissingValMask` (which derives from the stream attribute) is `.true.` and the array dimensions are appropriate. It applies the masking through a call  to `mpas_mask_field`.
5. `mpas_field_routines.F` contains `mpas_mask_assign_pointer` and `mpas_mask_field` for multiple mask and field dimension combinations.
6. Support was added for the `time_series_stats` analysis member through changes to `duplicate_field_array.inc` and assigning pointer to `missingValueMaskXd` in subroutines of `mpas_ocn_time_series_stats.F`. The latter needed to be located in dimension-specific routines so that the appropriate X values were used.
7. The `global_ocean` test cases had an implicit assumption that `normalVelocity` would be initialized to 0 that needed to be made explicit in `mpas_ocn_init_global_ocean.F` in order for the PR test suite to pass.

**To use this feature:**

1. If the variable you want to mask does not have `missing_value_mask="maskingArray"` in its entry in the Registry, add this line.
2. Add `useMissingValMask="true"` to the stream entry to which you would like to apply the masking.

[BFB]

### CODE REVIEW

**Comment by cbegeman** [comment]
See https://github.com/E3SM-Ocean-Discussion/E3SM/pull/25 for discussion.

**Comment by rljacob** [comment]
What does FWK mean?

**Comment by cbegeman** [comment]
MPAS-Framework. Now spelled out in the title.

**Comment by jonbob** [comment]
@cbegeman - just a reminder that the branch naming convention looks like:
<Github username>/<source code area or component>/<feature-description>
Even if it's on your fork, I think it needs to have your GitHub username. It may not be necessary to change it this time, just a friendly reminder....

File: components/mpas-framework/src/framework/duplicate_field_array.inc [hunk 1]
```diff
@@ -48,6 +48,8 @@
                dst_cursor % hasTimeDimension = src_cursor % hasTimeDimension
                dst_cursor % dimNames = src_cursor % dimNames
                dst_cursor % dimSizes = src_cursor % dimSizes
+	       dst_cursor % maskName = src_cursor % maskName
+	       dst_cursor % missingValue = src_cursor % missingValue
```
Comment (mark-petersen) [inline components/mpas-framework/src/framework/duplicate_field_array.inc:52]: Tabs should be changed to spaces here.

**Comment by mark-petersen** [comment]
@cbegeman thanks, this is a lot of work! This compiles as-is with both gnu debug and intel debug on badger, and passes the nightly test suite.

I'm trying to show that it is working. I see these lines in the ocean Registry:
```
2112         <var name="normalVelocity" type="real" dimensions="nVertLevels nEdges Time" units="m s^{-1}"
2113              description="horizontal velocity, normal component to an edge"
2114              missing_value="FILLVAL" missing_value_mask="edgeMask"
2115         />
2116         <var name="layerThickness" type="real" dimensions="nVertLevels nCells Time" units="m"
2117              description="layer thickness"
2118              missing_value="FILLVAL" missing_value_mask="cellMask"
```
So these two indeed have `missing_value` and `missing_value_mask`. On a little test run, the default Compass test `ocean/global_ocean/QU240/PHC/performance_test/forward` runs without error. 

 I then use `MissingValMask="true"` to the stream entry:
```
 22 <stream name="output"
 23         type="output"
 24         filename_template="output.nc"
 25         filename_interval="01-00-00_00:00:00"
 26         reference_time="0001-01-01_00:00:00"
 27         clobber_mode="truncate"
 28         precision="double"
 29         useMissingValMask="true"
 30         output_interval="0000_00:00:01">
 31
 32     <var_struct name="tracers"/>
 33     <var name="xtime"/>
 34     <var name="normalVelocity"/>
 35     <var name="layerThickness"/>
 36 </stream>
```
And I get the following error:
```
pwd
/lustre/scratch5/turquoise/mpeterse/runs/n/ocean_model_220830_bb84429e_ba_intel-mpi_19_openmp_master/ocean/global_ocean/QU240/PHC/performance_test/forward
(dev_compass_1.2.0-alpha.1) ba177:forward$ srun -n 4 ocean_model
forrtl: severe (408): fort: (2): Subscript #1 of the array MASKEDFIELD has value 2 which is greater than the upper bound of 1

Image              PC                Routine            Line        Source
libpnetcdf.so.4    00002ABF1D7080C9  for_emit_diagnost     Unknown  Unknown
ocean_model        0000000003592C08  mpas_field_routin        3426  mpas_field_routines.F
ocean_model        000000000345ED1B  mpas_io_streams_m        3776  mpas_io_streams.F
ocean_model        00000000034FFB54  mpas_stream_manag        3382  mpas_stream_manager.F
ocean_model        00000000034FAE7F  mpas_stream_manag        2819  mpas_stream_manager.F
ocean_model        0000000001730937  ocn_forward_mode_         712  mpas_ocn_forward_mode.F
ocean_model        0000000001F19AF8  ocn_core_mp_ocn_c         111  mpas_ocn_core.F
ocean_model        00000000004173F5  mpas_subdriver_mp         358  mpas_subdriver.F
ocean_model        0000000000412114  MAIN__                     20  mpas.F
ocean_model        0000000000412052  Unknown               Unknown  Unknown
libc-2.17.so       00002ABF1FE2C555  __libc_start_main     Unknown  Unknown
ocean_model        0000000000411F69  Unknown               Unknown  Unknown
```
which is in the new code.

**Comment by jonbob** [comment]
@cbegeman - preliminary testing in E3SM show that none of these changes result in non-BFB reports from cprnc as part of typical testing. @xylar had thought we should confirm that...

**Comment by cbegeman** [comment]
@mark-petersen Thanks for giving this a try. Your streams file seems identical to mine. Can you let me know the machine and compiler and I'll try to reproduce the error?

**Comment by xylar** [comment]
I tried re-testing this but have been stuck in the queue on Anvil all day.

**Comment by mark-petersen** [comment]
> Can you let me know the machine and compiler and I'll try to reproduce the error?
Badger, with intel 19 and debug:
```
(dev_compass_1.2.0-alpha.1) ba081:forward$ pwd
/lustre/scratch5/turquoise/mpeterse/runs/n/ocean_model_220830_bb84429e_ba_intel-mpi_19_openmp_master/ocean/global_ocean/QU240/PHC/performance_test/forward
(dev_compass_1.2.0-alpha.1) ba081:forward$ module list

Currently Loaded Modules:
  1) cmake/3.19.2   2) intel/19.0.4   3) intel-mpi/2019.4   4) mkl/2019.0.4
```
I am retesting your revisions now.

**Comment by mark-petersen** [comment]
Yes, that works now. Just to be clear, when I add `MissingValMask="true"` to the stream entry:
```
 22 <stream name="output"
 23         type="output"
 24         filename_template="output.nc"
 25         filename_interval="01-00-00_00:00:00"
 26         reference_time="0001-01-01_00:00:00"
 27         clobber_mode="truncate"
 28         precision="double"
 29         useMissingValMask="true"
 30         output_interval="0000_00:00:01">
 31
 32     <var_struct name="tracers"/>
 33     <var name="xtime"/>
 34     <var name="normalVelocity"/>
 35     <var name="layerThickness"/>
 36 </stream>
```
I now see the fill value:
```
pwd
/lustre/scratch5/turquoise/mpeterse/runs/n/ocean_model_220830_bb84429e_ba_intel-mpi_19_openmp_master/ocean/global_ocean/QU240/PHC/performance_test/forward
(dev_compass_1.2.0-alpha.1) ba081:forward$ ncdump -v normalVelocity output.nc |less
...
        double normalVelocity(Time, nEdges, nVertLevels) ;
                normalVelocity:units = "m s^-1" ;
                normalVelocity:long_name = "horizontal velocity, normal component to an edge" ;
                normalVelocity:missing_value_mask = "edgeMask" ;
                normalVelocity:_FillValue = 9.96920996838687e+36 ;
        double layerThickness(Time, nCells, nVertLevels) ;
                layerThickness:units = "m" ;
                layerThickness:long_name = "layer thickness" ;
                layerThickness:missing_value_mask = "cellMask" ;
                layerThickness:_FillValue = 9.96920996838687e+36 ;
...
 normalVelocity =
 ...
  0.0504185826930351, 0.0192160381013135, 0.00194918323427428,
    0.00294431925911961, 0.00317495330474025, _, _, _, _, _, _, _, _, _, _, _,
```
and ncdump can interpret the fill value and shows a blank (`_`).

When I take the `MissingValMask="true"`  out and rerun, those entries show up as zero, rather than blank:
```
 normalVelocity =
...
  0.0504185826930351, 0.0192160381013135, 0.00194918323427428,
    0.00294431925911961, 0.00317495330474025, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
```

Great job @cbegeman! Next I'll try some test runs with `MissingValMask="true"` in _every_ stream. @cbegeman and @xylar, can you think of any reason to not include this now in the streams files? Have you tested it on restart streams for both output and input? What happens if 'old' restart files has it off, but we turn it on in restart streams? Is there any effect on reading values in? I would expect not.

**Comment by xylar** [comment]
> Next I'll try some test runs with `MissingValMask="true"` in _every_ stream. @cbegeman and @xylar, can you think of any reason to not include this now in the streams files? 

This definitely can't be done in restart files.  It would lead to non-bit-for-bit behavior precisely for the reasons that we couldn't just put in missing values for invalid entries in these fields in the first place.

> Have you tested it on restart streams for both output and input? 

Yes, it doesn't work.  So restart streams definitely need to not have the `MissingValMask="true"` attribute.

> What happens if 'old' restart files has it off, but we turn it on in restart streams? Is there any effect on reading values in? I would expect not.

That not be a problem if missing values were handled as fill values in the code but it's a moot point here because restart files can't have missing values masked.

**Comment by xylar** [comment]
My `pr` tests on Anvil with Intel and Intel-MPI passed and were bit-for-bit expect:
```
Variable dThreshMLD not in /lcrc/group/e3sm/ac.xylar/compass_1.2/anvil/test_20220831/ocean_pr_fill_mask/ocean/global_ocean/QU240/PHC/analysis_test/forward/analysis_members/mixedLayerDepths.0001-01-01.nc.
Variable dThreshMLD not in /lcrc/group/e3sm/ac.xylar/compass_1.2/anvil/test_20220831/ocean_pr_e3sm_master/ocean/global_ocean/QU240/PHC/analysis_test/forward/analysis_members/mixedLayerDepths.0001-01-01.nc.
```

I believe this is unrelated to this PR.

**Comment by mark-petersen** [comment]
@xylar thanks for your answer on the restart files. Do you think we can add `MissingValMask="true"` to all other output streams with the code changes in this PR?

**Comment by xylar** [comment]
> Do you think we can add MissingValMask="true" to all other output streams with the code changes in this PR?

I would hold off for now.  I suggested keeping masking on for a few variables here just to make sure things worked as expected and to make it easy to show.  Since we only have a few test variables with masks so far, it could also wait for a follow-up PR that puts masks on more variables and also turns on the `MissingValMask="true"` attribute.  I think it makes sense for this PR to primarily focus on the framework.

I also don't want to add work for @cbegeman, who I think is ready to be done with this particular project and to move on to other development.

**Review by xylar — APPROVED** [review]
Approved based on my testing in https://github.com/E3SM-Ocean-Discussion/E3SM/pull/25 and bit-for-bit testing of the current branch using the compass `pr` test suite.

**Comment by cbegeman** [comment]
@mark-petersen Thanks for giving that fix a try! I'm glad it worked out. 

@xylar Thanks for your retesting and for explaining our thinking on the PR scope. 

I'd be happy to work on the subsequent PR that adds mask attributes to more variables and turns on masking in some streams when my schedule frees up a bit more in a few months.

**Review by mark-petersen — APPROVED** [review]
@cbegeman with this last commit, `useMissingValMask="true"` appears in every output stream that comes from the default standalone streams, i.e. the ones used by compass. I compiled and ran with this, and it passes the nightly suite for both intel debug and gnu debug. This includes the QU240 analysis test, where almost all the analysis members are on, and the extra flag was included.

Note that E3SM stream files were not changed in this PR. I think that is safer, in case there is any i/o or MPAS-Analysis problem. But we can do that in a future PR, as well as add `missing_value="FILLVAL"` to all the other variables.

**Comment by cbegeman** [comment]
@mark-petersen Great! Thanks for trying that out in those output streams.

@mark-petersen and @xylar Does it seem appropriate to you to include Mark's commit https://github.com/E3SM-Project/E3SM/pull/5154/commits/10f01f692d574071cbc2f319a5bf8d11b5bffa5f in this PR? I think we had originally talked about including this in a subsequent PR so that the feature isn't fully rolled out until the masking is applied to all relevant variables. If we apply masking now for MPAS Standalone output streams, it might be a bit confusing to users that masking is applied to some variables and not others.

**Comment by xylar** [comment]
@mark-petersen, see my comment from yesterday: https://github.com/E3SM-Project/E3SM/pull/5154#issuecomment-1233125199

I really think we want to keep this focused on MPAS-Framework changes and bring in your commit 10f01f6 in a separate PR.

**Comment by mark-petersen** [comment]
I thought of that, but there were already changes to the ocean `Registry.xml` and tracer registry files, so this PR already has framework and ocean changes together. Applying the stream flag now has the advantage that the prognostic variables, at least, are written with correct masks in the stand-alone output. But I can go either way.

**Comment by xylar** [comment]
@mark-petersen, as I said, I hesitated at first to have any MPAS-Ocean changes here.  But in my other recent MPAS-Framework PRs, @matthewhoffman has found it dissatisfying to not have a demonstration of what the framework change is meant to do in the components.  For that reason, I thought having 3 or 4 variables getting masked would be a good demonstration of principle and I suggested that @cbegeman leave those in.

But I think it's a big step from a few demonstration variables turning on masking in several dozen ocean registry files.  For one, it makes the review quite a bit more challenging for non-ocean reviewers.  And it isn't clear why similar changes were not made in other components. One could reasonably ask why we don't turn masking on for a few variables in each component but I think the reason is simply that we wanted just a few to demonstrate the concept.

I must firmly request that 10f01f6 be removed and included as part of a later PR that is more focused on MPAS-Ocean.

**Comment by mark-petersen** [comment]
OK, let's do that. @cbegeman can you remove my last commit?

**Comment by cbegeman** [comment]
@mark-petersen No problem! I'll include that commit in the next PR.

**Review by matthewhoffman — COMMENTED** [review]
@cbegeman , this looks like a lot of work, especially and the repetitive interfaces!  I have a couple questions, but otherwise I have no concerns.  I would like to try to test this on some MALI runs, and I will try to do that this week yet.

File: components/mpas-framework/src/framework/mpas_field_types.inc [hunk 2]
```diff
@@ -224,6 +242,7 @@
       integer :: defaultValue
       integer :: missingValue
       integer, dimension(3) :: dimSizes
+      character (len=StrKIND) :: maskName
```
[the remark this answers is not in the document]
Comment (cbegeman) [reply components/mpas-framework/src/framework/mpas_field_types.inc:245]: Oh, actually it is a problem here (and maybe elsewhere) if the fields don't have the `maskName` attribute:
https://github.com/E3SM-Project/E3SM/blob/44f67eff2e2899b23baa79f085e037ad80aab2ad/components/mpas-framework/src/framework/duplicate_field_array.inc#L51
I think the framework generally has the same attribute for all fields, and it is either associated or not.

[the remark this answers is not in the document]
Comment (cbegeman) [reply components/mpas-framework/src/framework/mpas_field_types.inc:245]: I'm not having too much success with that pointer route (problems in block creator). If this isn't very important to you, I suggest we leave it as-is.

[the remark this answers is not in the document]
Comment (xylar) [reply components/mpas-framework/src/framework/mpas_field_types.inc:245]: Looking at that `duplicate` code, I think @cbegeman is right that there are too many built-in assumptions that all fields have the same attributes.

File: components/mpas-framework/src/framework/mpas_field_routines.F [hunk 3]
```diff
@@ -2858,6 +2888,1172 @@ subroutine mpas_allocate_mold_1dchar(dst, src)!{{{
...
+!> Masks the source field using a masking array.
+!
+!-----------------------------------------------------------------------
+   subroutine mpas_mask_field1d_mask1d_real(field, mask, maskedField) !{{{
+
+      implicit none
+
+      type (field1DReal), intent(in), pointer :: field
+      type (field1DInteger), intent(in), pointer :: mask
+      real (kind=RKIND), dimension(:), intent(inout), pointer :: maskedField
```
[the remark this answers is not in the document]
Comment (cbegeman) [reply components/mpas-framework/src/framework/mpas_field_routines.F]: Addressed by https://github.com/E3SM-Project/E3SM/pull/5154/commits/a9a6fb7f39a98f83f8b756aa80e94f5262d0c187. I changed `maskedField` to `maskedArray` and `maskArray` to `maskingArray` for clarity.

File: components/mpas-framework/src/framework/mpas_field_types.inc (hunk 2 above)
[the remark this answers is not in the document]
Comment (cbegeman) [reply components/mpas-framework/src/framework/mpas_field_types.inc:245]: @matthewhoffman Just wanted to follow up and see if keeping the `maskName` attribute for all fields is alright with you.

**Review by akturner — APPROVED** [review]
Passed MPAS-Seaice standalone tests

File: components/mpas-framework/src/framework/mpas_field_types.inc (hunk 2 above)
[the remark this answers is not in the document]
Comment (cbegeman) [reply components/mpas-framework/src/framework/mpas_field_types.inc:245]: @matthewhoffman Reminding you to take a look at this when you have a chance. Thanks!

[the remark this answers is not in the document]
Comment (matthewhoffman) [reply components/mpas-framework/src/framework/mpas_field_types.inc:245]: @cbegeman , sorry for the slow reply!  Thanks for looking into it, and I agree with your conclusion that it is better to let all fields have the same attributes than try to find a way to relax the assumption.  Keeping `maskName` for all fields is fine with me!

**Review by matthewhoffman — APPROVED** [review]
@cbegeman , thanks for responding to my suggestions and making the cosmetic changes.

**Comment by jonbob** [comment]
passes sanity testing with:
* SMS_D_Ld3.T62_oQU120.CMPASO-IAF.chrysalis_intel
* ERS.ne11_oQU240.WCYCL1850NS.chrysalis_intel

merged to next

**Comment by cbegeman** [comment]
Thanks for testing, @jonbob!

**Comment by jonbob** [comment]
This PR is responsible for the FAILs on compy with pgi, using any mpas components. Those errors appear in the e3sm logs and look like:
````
 80: 0: ALLOCATE: 4359484439294640128 bytes requested; not enough memory
````
As far as I can tell, it happens during the I/O phase after the simulation has effectively ended. There is no traceback information, and the tests are successful in debug mode. I ran some C-cases by hand and found that for mpas-ocean the failure may be related to the highFrequency output stream -- that test completes with timeSeriesStats output enabled but config_am_highfrequencyoutput_enable = .false.

**Comment by cbegeman** [comment]
> This PR is responsible for the FAILs on compy with pgi, using any mpas components. Those errors appear in the e3sm logs and look like:
> 
> ```
>  80: 0: ALLOCATE: 4359484439294640128 bytes requested; not enough memory
> ```
> 
> As far as I can tell, it happens during the I/O phase after the simulation has effectively ended. There is no traceback information, and the tests are successful in debug mode. I ran some C-cases by hand and found that for mpas-ocean the failure may be related to the highFrequency output stream -- that test completes with timeSeriesStats output enabled but config_am_highfrequencyoutput_enable = .false.

@akturner Any ideas about what might be causing the issue?

**Comment by xylar** [comment]
This would likely happen when calling `allocate` with a garbage value as the size.  I would guess the only way to debug it, given what @jonbob said is to put print statements in the code to try to find where it's happening.  Not very fun!

**Comment by jonbob** [comment]
@xylar Agreed, both that it's the most reasonable approach -- and no fun!

**Comment by cbegeman** [comment]
Well, there's only one allocate statement that I added (repeated for a bunch of dims) https://github.com/E3SM-Project/E3SM/blob/673106df1d9159c1c8a30ec6ce2170ed4cc3a8b6/components/mpas-framework/src/framework/mpas_io_streams.F#L3600, so this maybe points to instances where `ownedSize` is ill-defined.

**Comment by xylar** [comment]
If `ownedSize` is the problem, it seems like that would point to the dimensions defining a field being deallocated before the final writing out of that field, which would seem like a pretty major bug (obviously not introduced by this PR).  Does it seem like we should just wait to see what PGI results look like on machines where we actually plan to use PGI?  And debug further if those results are also showing this problem?

**Comment by rljacob** [comment]
Instead of adding print statements you'll have to later take out, add code to capture the error status of each allocate and then die if its non zero.

Every allocate statement in MCT looks something like this:
```
 allocate(Rout%ip1(proc)%pi(Rout%locsize(proc)*numi),stat=ier)
 if(ier/=0) call die(myname_,'allocate(Rout%ip1%pi)',ier)
```

**Comment by cbegeman** [comment]
Thanks for the tip, @rljacob!

**Comment by jonbob** [comment]
@rljacob -- I added the allocate error status checks, pretty similar to what we do in MCT, and it now runs OK. Does that make any sense to you? Do we just chalk it up to old PGI madness?

**Comment by rljacob** [comment]
Yes I suppose we can say that's the old PGI compiler for ya.

**Comment by jonbob** [comment]
Well, we'll plan to re-merge tomorrow with a fix... @rljacob, thanks for your help

**Comment by cbegeman** [comment]
@jonbob I've added your changes and rebased onto master. I also rechecked that the masking feature works. Thanks for all your work on this!

**Comment by jonbob** [comment]
That looks better, thanks!

re-merged to next to fix problems on compy-pgi

**Comment by jonbob** [comment]
The updates cleared testing on compy with pgi

Merged to master

**Comment by xylar** [comment]
Yay! Thanks @jonbob and many thanks to @cbegeman for the hard work on this!  I know it turned out to be a way bigger project than you anticipated but MPAS output will be much the better for it.

**Pull request merged**